src/context/defrag.rs

```diff
@@ -0,0 +1,226 @@
+use std::alloc::Layout;
+
+use crate::{
+    raw, Context, RedisModule_DefragAlloc, RedisModule_DefragCursorGet,
+    RedisModule_DefragCursorSet, RedisModule_DefragRedisModuleString, RedisModule_DefragShouldStop,
+    RedisString, Status,
+};
+use crate::{RedisError, RedisLockIndicator};
+use linkme::distributed_slice;
+
+pub struct DefragContext {
+    defrag_ctx: *mut raw::RedisModuleDefragCtx,
+}
+
+/// Having a [DefragContext] is indication that we are
+/// currently holding the Redis GIL, this is why it is safe to
+/// implement a [RedisLockIndicator] for [DefragContext].
+unsafe impl RedisLockIndicator for DefragContext {}
+
+impl DefragContext {
+    /// Creates a new [`DefragContext`] from a poiter to [`raw::RedisModuleDefragCtx`].
+    /// The function is considered unsafe because the provided pointer
+    /// must be a valid pointer to [`raw::RedisModuleDefragCtx`], and the Redis GIL must be held.
+    /// The function is exposed for users that wants to implement the defrag function
+    /// on their module datatype, they can use this function to create [`DefragContext`]
+    /// that can be used in a safely manner.
+    /// Notice that the returned [`DefragContext`] borrows the pointer to [`raw::RedisModuleDefragCtx`]
+    /// so it can not outlive it (this means that it should not be used once the defrag callback ends).
+    pub unsafe fn new(defrag_ctx: *mut raw::RedisModuleDefragCtx) -> DefragContext {
+        DefragContext { defrag_ctx }
+    }
+
+    /// When the data type defrag callback iterates complex structures, this
+    /// function should be called periodically. A [`false`] return
+    /// indicates the callback may continue its work. A [`true`]
+    /// indicates it should stop.
+    ///
+    /// When stopped, the callback may use [`Self::set_cursor`] to store its
+    /// position so it can later use [`Self::get_cursor`] to resume defragging.
+    ///
+    /// When stopped and more work is left to be done, the callback should
+    /// return 1. Otherwise, it should return 0.
+    ///
+    /// NOTE: Modules should consider the frequency in which this function is called,
+    /// so it generally makes sense to do small batches of work in between calls.
+    pub fn should_stop(&self) -> bool {
+        let should_stop = unsafe {
+            RedisModule_DefragShouldStop.expect("RedisModule_DefragShouldStop is NULL")(
```

Comment:
`RedisModule_DefragShouldStop.expect("RedisModule_DefragShouldStop is available")`

According to the Rust docs,
> Recommended Message Style
> We recommend that expect messages are used to describe the reason you expect the Option should be Some.